CI: Test_getOldLeases() failed in Travis

#9972 is failing due to a unit test failure in the newly introduced unit tests from #9982:

https://travis-ci.com/cilium/cilium/builds/146829104

```
--- FAIL: Test_getOldestLeases (0.00s)
751    --- PASS: Test_getOldestLeases/test-1 (0.00s)
752    --- FAIL: Test_getOldestLeases/test-2 (0.00s)
753        kvstore_watchdog_test.go:85: getOldestLeases() = map[foo/bar/1:{[] 1 1 } foo/bar/2:{[] 2 2 } foo/baz/4:{[] 4 4 }], want map[foo/bar/1:{[] 1 1 } foo/baz/4:{[] 4 4 }]
754=== RUN   Test_getPath
755=== RUN   Test_getPath/test-1
756=== RUN   Test_getPath/test-2
757--- PASS: Test_getPath (0.00s)
758    --- PASS: Test_getPath/test-1 (0.00s)
759    --- PASS: Test_getPath/test-2 (0.00s)
760FAIL
761coverage: 2.0% of statements in ./...
762FAIL	github.com/cilium/cilium/operator	0.067s
763FAIL
```

--- a/operator/kvstore_watchdog.go
+++ b/operator/kvstore_watchdog.go
@@ -39,16 +39,27 @@ func keyPathFromLockPath(k string) string {
 
 // getOldestLeases returns the value that has the smaller revision for each
 // 'path'. A 'path' shares the same common prefix for different locks.
-func getOldestLeases(m map[string]kvstore.Value) map[string]kvstore.Value {
-	oldestLeases := map[string]kvstore.Value{}
-	oldestPaths := map[string]kvstore.Value{}
-	for k, v := range m {
-		commonPath := keyPathFromLockPath(k)
-		oldestLease, ok := oldestPaths[commonPath]
-		if !ok || v.ModRevision < oldestLease.ModRevision {
-			oldestPaths[commonPath] = v
-			oldestLeases[k] = v
+func getOldestLeases(lockPaths map[string]kvstore.Value) map[string]kvstore.Value {
+	type LockValue struct {
+		kvstore.Value
+		keyPath string
+	}
+	oldestPaths := map[string]LockValue{}
+	for lockPath, v := range lockPaths {
+		keyPath := keyPathFromLockPath(lockPath)
+		oldestKeyPath, ok := oldestPaths[keyPath]
+		if !ok || v.ModRevision < oldestKeyPath.ModRevision {
+			// Store the oldest common path
+			oldestPaths[keyPath] = LockValue{
+				keyPath: lockPath,
+				Value:   v,
+			}
 		}
+	}
+	oldestLeases := map[string]kvstore.Value{}
+	for _, v := range oldestPaths {
+		// Retrieve the oldest lock path
+		oldestLeases[v.keyPath] = v.Value
 	}
 	return oldestLeases
 }

--- a/operator/kvstore_watchdog_test.go
+++ b/operator/kvstore_watchdog_test.go
@@ -58,10 +58,25 @@ func Test_getOldestLeases(t *testing.T) {
 						ModRevision: 3,
 						LeaseID:     3,
 					},
-					"foo/baz/4": {
+					"foo/bar/4": {
 						Data:        nil,
 						ModRevision: 4,
 						LeaseID:     4,
+					},
+					"foo/bar/5": {
+						Data:        nil,
+						ModRevision: 5,
+						LeaseID:     5,
+					},
+					"foo/baz/6": {
+						Data:        nil,
+						ModRevision: 6,
+						LeaseID:     6,
+					},
+					"foo/bbz/7": {
+						Data:        nil,
+						ModRevision: 3,
+						LeaseID:     3,
 					},
 				},
 			},
@@ -71,10 +86,15 @@ func Test_getOldestLeases(t *testing.T) {
 					ModRevision: 1,
 					LeaseID:     1,
 				},
-				"foo/baz/4": {
+				"foo/baz/6": {
 					Data:        nil,
-					ModRevision: 4,
-					LeaseID:     4,
+					ModRevision: 6,
+					LeaseID:     6,
+				},
+				"foo/bbz/7": {
+					Data:        nil,
+					ModRevision: 3,
+					LeaseID:     3,
 				},
 			},
 		},

--- a/pkg/kvstore/allocator/allocator_test.go
+++ b/pkg/kvstore/allocator/allocator_test.go
@@ -125,6 +125,12 @@ func (s *AllocatorSuite) TestRunLocksGC(c *C) {
 	allocator, err := allocator.NewAllocator(TestAllocatorKey(""), backend1, allocator.WithMax(maxID), allocator.WithoutGC())
 	c.Assert(err, IsNil)
 	shortKey := TestAllocatorKey("1;")
+
+	staleLocks := map[string]kvstore.Value{}
+	staleLocks, err = allocator.RunLocksGC(context.Background(), staleLocks)
+	c.Assert(err, IsNil)
+	c.Assert(len(staleLocks), Equals, 0)
+
 	var (
 		lock1, lock2 kvstore.KVLocker
 		gotLock1     = make(chan struct{})
@@ -161,10 +167,6 @@ func (s *AllocatorSuite) TestRunLocksGC(c *C) {
 		c.Assert(err, IsNil)
 		close(gotLock2)
 	}()
-	staleLocks := map[string]kvstore.Value{}
-	staleLocks, err = allocator.RunLocksGC(context.Background(), staleLocks)
-	c.Assert(err, IsNil)
-	c.Assert(len(staleLocks), Equals, 0)
 
 	// Wait until lock1 is gotten.
 	c.Assert(testutils.WaitUntil(func() bool {

--- a/pkg/kvstore/allocator/allocator_test.go
+++ b/pkg/kvstore/allocator/allocator_test.go
@@ -157,7 +157,7 @@ func (s *AllocatorSuite) TestRunLocksGC(c *C) {
 				nil,
 			)
 		}
-		lock2, err = client.LockPath(context.Background(), allocatorName+"/locks/"+shortKey.GetKey())
+		lock2, err = client.LockPath(context.Background(), allocatorName+"/locks/"+kvstore.Encode(shortKey.GetKey()))
 		c.Assert(err, IsNil)
 		close(gotLock2)
 	}()
@@ -184,7 +184,14 @@ func (s *AllocatorSuite) TestRunLocksGC(c *C) {
 	// Check which locks are stale, it should be lock1 and lock2
 	staleLocks, err = allocator.RunLocksGC(context.Background(), staleLocks)
 	c.Assert(err, IsNil)
-	c.Assert(len(staleLocks), Equals, 2)
+	switch s.backend {
+	case "consul":
+		// Contrary to etcd, consul does not create a lock in the kvstore
+		// if a lock is already being held.
+		c.Assert(len(staleLocks), Equals, 1)
+	case "etcd":
+		c.Assert(len(staleLocks), Equals, 2)
+	}
 
 	var (
 		oldestRev     uint64

